internal/restore/pvc_action.go

```diff
@@ -189,19 +191,81 @@ func (p *PVCRestoreItemAction) Progress(operationID string, restore *velerov1api
+	})
+
+	dataDownload, err := getDataDownload(context.Background(), restore.Namespace, operationID, p.VeleroClient)
+	if err != nil {
+		logger.Errorf("fail to get DataDownload: %s", err.Error())
+		return progress, err
+	}
+	if dataDownload.Status.Phase == velerov2alpha1.DataDownloadPhaseNew ||
+		dataDownload.Status.Phase == "" {
+		logger.Debugf("DataDownload is still not processed yet. Skip progress update.")
+		return progress, nil
+	}
+
+	progress.Description = string(dataDownload.Status.Phase)
+	progress.OperationUnits = "Bytes"
+	progress.NCompleted = dataDownload.Status.Progress.BytesDone
+	progress.NTotal = dataDownload.Status.Progress.TotalBytes
+
+	if dataDownload.Status.StartTimestamp != nil {
+		progress.Started = dataDownload.Status.StartTimestamp.Time
+	}
+
+	if dataDownload.Status.CompletionTimestamp != nil {
+		progress.Updated = dataDownload.Status.CompletionTimestamp.Time
+	}
+
+	if dataDownload.Status.Phase == velerov2alpha1.DataDownloadPhaseCompleted {
+		progress.Completed = true
+	} else if dataDownload.Status.Phase == velerov2alpha1.DataDownloadPhaseCanceled {
+		progress.Completed = true
+		progress.Err = fmt.Sprintf("DataDownload is canceled")
+	} else if dataDownload.Status.Phase == velerov2alpha1.DataDownloadPhaseFailed {
+		progress.Completed = true
+		progress.Err = dataDownload.Status.Message
+	}
 
 	return progress, nil
 }
 
 func (p *PVCRestoreItemAction) Cancel(operationID string, restore *velerov1api.Restore) error {
-	return nil
+	if operationID == "" {
+		return riav2.InvalidOperationIDError(operationID)
+	}
+	logger := p.Log.WithFields(logrus.Fields{
+		"Action":      "PVCRestoreItemAction",
+		"OperationID": operationID,
+		"Namespace":   restore.Namespace,
+	})
+
+	dataDownload, err := getDataDownload(context.Background(), restore.Namespace, operationID, p.VeleroClient)
+	if err != nil {
+		logger.Errorf("fail to get DataDownload: %s", err.Error())
+		return err
+	}
+
+	err = cancelDataDownload(context.Background(), p.VeleroClient, dataDownload)
+	if err != nil {
+		logger.Errorf("fail to cancel DataDownload %s: %s", dataDownload.Name, err.Error())
+	}
+	return err
+}
+
+func (p *PVCRestoreItemAction) AreAdditionalItemsReady(additionalItems []velero.ResourceIdentifier, restore *velerov1api.Restore) (bool, error) {
+	return true, nil
 }
 
 func getDataUploadResult(ctx context.Context, restore *velerov1api.Restore, pvc *corev1api.PersistentVolumeClaim,
 	kubeClient kubernetes.Interface) (*velerov2alpha1.DataUploadResult, error) {
 	cmList, err := kubeClient.CoreV1().ConfigMaps(restore.Namespace).List(ctx, metav1.ListOptions{
-		LabelSelector: fmt.Sprintf("%s=%s,%s=%s", velerov1api.BackupNameLabel, label.GetValidName(restore.Spec.BackupName),
-			util.PVCNamespaceNameLabel, pvc.Namespace+"/"+pvc.Name),
+		LabelSelector: fmt.Sprintf("%s=%s,%s=%s", util.PVCNamespaceNameLabel, pvc.Namespace+"."+pvc.Name,
+			velerov1api.BackupNameLabel, label.GetValidName(restore.Spec.BackupName)),
```

Comment:
As we discussed, this label cannot uniquely identify a cm, since two restores of the same PVC from the same backup will get the same label. So probably, we need to add the restore name to the label?

Reply:
Restore UID looks better. Is the backup name label used in other places?
How about changing like this?
```
LabelSelector: fmt.Sprintf("%s=%s,%s=%s", util.PVCNamespaceNameLabel, pvc.Namespace+"."+pvc.Name,
			velerov1api.RestoreUIDLabel, label.GetValidName(restore.UID))
```

Reply:
OK, so the creation counterpart on Velero side should also be changed.